examples/digital_fingerprinting/production/morpheus/dfp/stages/dfp_file_to_df.py

```diff
@@ -0,0 +1,257 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import json
+import logging
+import multiprocessing as mp
+import os
+import time
+import typing
+from functools import partial
+
+import fsspec
+import pandas as pd
+import srf
+from srf.core import operators as ops
+
+import dask
+from dask.distributed import Client
+from dask.distributed import LocalCluster
+
+import cudf
+
+from morpheus._lib.file_types import FileTypes
+from morpheus.config import Config
+from morpheus.io.deserializers import read_file_to_df
+from morpheus.pipeline.single_port_stage import SinglePortStage
+from morpheus.pipeline.stream_pair import StreamPair
+
+from ..utils.column_info import DataFrameInputSchema
+from ..utils.column_info import process_dataframe
+
+logger = logging.getLogger("morpheus.{}".format(__name__))
+
+
+def _single_object_to_dataframe(file_object: fsspec.core.OpenFile,
+                                schema: DataFrameInputSchema,
+                                file_type: FileTypes,
+                                filter_null: bool,
+                                parser_kwargs: dict):
+
+    retries = 0
+    s3_df = None
+    while (retries < 2):
+        try:
+            with file_object as f:
+                s3_df = read_file_to_df(f,
+                                        file_type,
+                                        filter_nulls=filter_null,
+                                        df_type="pandas",
+                                        parser_kwargs=parser_kwargs)
+
+            break
+        except Exception as e:
+            if (retries < 2):
+                logger.warning("Refreshing S3 credentials")
+                # cred_refresh()
+                retries += 1
+            else:
+                raise e
+
+    # Run the pre-processing before returning
+    if (s3_df is None):
+        return s3_df
+
+    s3_df = process_dataframe(df_in=s3_df, input_schema=schema)
+
+    return s3_df
+
+
+class DFPFileToDataFrameStage(SinglePortStage):
+
+    def __init__(self,
+                 c: Config,
+                 schema: DataFrameInputSchema,
+                 filter_null: bool = True,
+                 file_type: FileTypes = FileTypes.Auto,
+                 parser_kwargs: dict = None,
+                 cache_dir: str = "./.cache/dfp"):
+        super().__init__(c)
+
+        self._schema = schema
+
+        self._batch_size = 10
+        self._batch_cache = []
+        self._file_type = file_type
+        self._filter_null = filter_null
+        self._parser_kwargs = {} if parser_kwargs is None else parser_kwargs
+        self._cache_dir = os.path.join(cache_dir, "file_cache")
+
+        self._dask_cluster: Client = None
+
+        self._download_method: typing.Literal["single_thread", "multiprocess", "dask",
+                                              "dask_thread"] = os.environ.get("FILE_DOWNLOAD_TYPE", "dask_thread")
```

Comment:
We might want to consider renaming the environment variable to something like `MORPHEUS_FILE_DOWNLOAD_TYPE` to avoid potential clashes with other systems.

Reply:
Done